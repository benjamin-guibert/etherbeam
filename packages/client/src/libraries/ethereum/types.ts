import { BigNumber } from 'ethers'

export const WETH_HASH = '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2'
export const USDC_HASH = '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48'

export enum AddressType {
  Unknown = 'unknown',
  Wallet = 'wallet',
  Token = 'token',
  Contract = 'contract',
  Transaction = 'transaction',
}

export enum TransactionMethodParameterType {
  Unknown = 'unknown',
  UnsignedInteger = 'unsigned_integer',
  Address = 'address',
  Addresses = 'addresses',
}

export enum TransactionStatus {
  Pending = 'pending',
  Mined = 'mined',
  Validated = 'validated',
  Cancelled = 'cancelled',
}

export enum TransactionActionType {
  Unknown = 'unknown',
  Approval = 'approval',
  Transfer = 'transfer',
  Swap = 'swap',
}

export enum TransactionActionDirection {
  Buy = 'buy',
  Sell = 'sell',
}

export interface TransactionMethodParameter {
  name: string
  type: TransactionMethodParameterType
  rawType: string
  value: TransactionMethodValueType
}

export enum HistoryTime {
  Hour = 'h',
  Day = 'd',
  Week = 'w',
  Month = 'm',
  Year = 'y',
}

export interface TokenPriceHistory {
  time: HistoryTime
  price: BigNumber
  ratio: number
}

export interface Token extends Address {
  name: string
  description?: string
  symbol: string
  decimals: number
  chartPair?: string
  price?: BigNumber
  priceHistory: TokenPriceHistory[]
  website?: string
  whitepaper?: string
  github?: string
  linkedin?: string
  facebook?: string
  reddit?: string
  twitter?: string
  telegram?: string
  discord?: string
  actions?: TransactionAction[]
}

export interface Address {
  hash: string
  url: string
  type: AddressType
  label?: string
}

export interface Transaction {
  blockNumber?: number
  dateTime: Date
  address: Address
  status: TransactionStatus
  from?: Address
  to?: Address
  value?: BigNumber
  gas: {
    unitPrice?: BigNumber
    limit?: number
    used?: number
    ratio?: number
    price?: BigNumber
  }
  action?: TransactionMethod
  logs?: TransactionMethod[]
}

export interface TransactionMethod {
  contract: Address
  name: string
  parameters: TransactionMethodParameter[]
}

export type TransactionMethodValueType = string | BigNumber | Address | Address[]

export interface TransactionAction {
  index: number
  transaction: Transaction
  type: TransactionActionType
  direction: TransactionActionDirection
  holder: Address
  fromAddress?: Address | Token
  fromAmount?: BigNumber
  toAddress?: Address | Token
  toAmount?: BigNumber
}
