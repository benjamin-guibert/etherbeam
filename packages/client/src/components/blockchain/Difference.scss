@use '../../styles/resources/fonts';
@use '../../styles/resources/sizes';

.my-difference-prefix {
  font-weight: bold;
  padding-right: sizes.$space-s;
}
