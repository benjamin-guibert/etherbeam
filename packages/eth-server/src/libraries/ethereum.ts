import { getLogger } from './logger'
import fs from 'fs'
import Web3 from 'web3'
import { isTransactionWatched, sleep, tryAction } from './helpers'
import { Contract, ContractType, Token, Transaction, WETH_HASH } from './types'
import {
  fetchContracts,
  fetchTokens,
  fetchTransactionsToUpdate,
  saveTransaction,
  saveTransactions,
  updateTokenPrice,
} from './server'
import { BlockHeader } from 'web3-eth'
import { Subscription } from 'web3-core-subscriptions'
import { fetchBlockTransactions, fetchPendingTransaction, fetchTransactionReceipt } from './web3'
import { BaseProvider, InfuraProvider, IpcProvider, JsonRpcProvider } from '@ethersproject/providers'
import { fetchTokenPrice } from './uniswap'

const logger = getLogger()

const ETH_IPC_FILE_PATH = 'geth/geth.ipc'
const RETRY_WAIT = 10 * 1000
const MAX_RETRIES = 10
const TRANSACTION_RECEIPTS_FETCHING_INTERVAL = 10 * 1000
const TOKEN_PRICE_FETCHING_INTERVAL = 30 * 1000

export interface EthereumData {
  web3: Web3
  web3Pool: Web3
  ethersProvider: BaseProvider
  contracts: Contract[]
  newBlocksSubscription: Subscription<BlockHeader>
  fetchingTransactionReceipts: boolean
  fetchingTokenPrices: boolean
  pendingTransactionsSubscription: Subscription<string>
}

const initializeProvider = (ipc: boolean, host: string): { web3: Web3; ethers: BaseProvider } => {
  logger.info(`Provider (IPC: ${ipc}, host: '${host}')...`)

  if (ipc) {
    try {
      if (fs.existsSync(ETH_IPC_FILE_PATH)) {
        logger.info(`IPC file found at '${ETH_IPC_FILE_PATH}'.`)
        return {
          web3: new Web3(new Web3.providers.IpcProvider(ETH_IPC_FILE_PATH, require('net'))),
          ethers: new IpcProvider(ETH_IPC_FILE_PATH),
        }
      } else {
        logger.info(`No IPC file at '${ETH_IPC_FILE_PATH}'.`)
      }
    } catch (error) {
      logger.error(`Error while looking for IPC file at '${ETH_IPC_FILE_PATH}'.`)
      logger.error(error.message)
      logger.error(error.stack)
    }
  }

  const providerHost = host?.toLowerCase()

  if (!providerHost) {
    logger.warn('Provider host is missing.')
    return {
      web3: new Web3(),
      ethers: null,
    }
  }

  const providers = { web3: null, ethers: null }

  if (providerHost.includes('infura')) {
    providers.ethers = new InfuraProvider(null, providerHost.split('/').pop())
  } else {
    providers.ethers = new JsonRpcProvider(host)
  }

  if (providerHost.startsWith('http')) {
    logger.info(`HTTP provider: '${host}'.`)
    providers.web3 = new Web3(new Web3.providers.HttpProvider(host))
  } else if (providerHost.startsWith('ws')) {
    logger.info(`Web Socket provider: '${host}'.`)
    providers.web3 = new Web3(new Web3.providers.WebsocketProvider(host))
  } else {
    providers.web3 = new Web3()
    logger.warn('No provider found.')
  }

  return providers
}

const initializeProviders = (data: EthereumData): void => {
  const { ETH_PROVIDER_IPC, ETH_PROVIDER, ETH_POOL_PROVIDER_IPC, ETH_POOL_PROVIDER } = process.env
  logger.info('Initializing base provider...')
  const { web3, ethers } = initializeProvider(ETH_PROVIDER_IPC == '1', ETH_PROVIDER)
  data.web3 = web3
  data.ethersProvider = ethers
  logger.info('Initializing pool provider...')
  data.web3Pool = initializeProvider(ETH_POOL_PROVIDER_IPC == '1', ETH_POOL_PROVIDER).web3
}

const initializeContracts = async (data: EthereumData): Promise<void> => {
  data.contracts = (await fetchContracts()).filter((contract) => contract.type == ContractType.Contract)
  data.contracts.push(...(await fetchTokens()))

  data.contracts.forEach((contract) => {
    const { hash, abi } = contract
    contract.web3 = new data.web3.eth.Contract(JSON.parse(abi), hash)
  })

  logger.info(`${data.contracts.length} contracts initialized.`)
}

export const useEthereum = async (): Promise<EthereumData> => {
  const data = {
    web3: undefined,
    web3Pool: undefined,
    ethersProvider: undefined,
    contracts: undefined,
    newBlocksSubscription: null,
    fetchingTransactionReceipts: false,
    fetchingTokenPrices: false,
    pendingTransactionsSubscription: null,
  }

  initializeProviders(data)
  await initializeContracts(data)

  return data
}

const tryFetchBlockTransactions = async (number: number, data: EthereumData): Promise<Transaction[]> => {
  return await tryAction<Transaction[]>(
    () => fetchBlockTransactions(number, data),
    RETRY_WAIT,
    MAX_RETRIES,
    (error) => {
      logger.error(`Error while fetching block #${number}.`)
      logger.error(error.message)
      logger.error(error.stack)
    },
    () => logger.info(`Block #${number} not available yet.`)
  )
}

const trySaveTransactions = async (transactions: Transaction[]): Promise<boolean> => {
  const result = await tryAction<boolean>(
    async () => {
      await saveTransactions(transactions)
      return true
    },
    RETRY_WAIT,
    MAX_RETRIES,
    (error) => {
      logger.error(`Error while saving ${transactions.length} transaction(s).`)
      logger.error(error.message)
      logger.error(error.stack)
    }
  )

  return !!result
}

const onNewBlock = async (blockHeader: BlockHeader, data: EthereumData): Promise<void> => {
  logger.debug(`New block header received: #${blockHeader.number}.`)

  const transactions = await tryFetchBlockTransactions(blockHeader.number, data)

  if (!transactions) return

  const filteredTransactions = transactions.filter((transaction) => isTransactionWatched(transaction, data.contracts))

  if (!filteredTransactions.length) return

  const result = await trySaveTransactions(filteredTransactions)

  if (result)
    logger.info(
      `${filteredTransactions.length}/${transactions.length} transaction(s) added from block #${blockHeader.number}.`
    )
}

export const subscribeToNewBlocks = (data: EthereumData): void => {
  if (data.newBlocksSubscription) return

  data.newBlocksSubscription = data.web3.eth
    .subscribe('newBlockHeaders')
    .on('connected', () => logger.info('Subscribed to new blocks.'))
    .on('data', (blockHeader) => onNewBlock(blockHeader, data))
    .on('error', (error) => {
      logger.error('Error from new blocks subscription.')
      logger.error(error.message)
      logger.error(error.stack)
    })
}

export const unsubscribeFromNewBlocks = (data: EthereumData): void => {
  if (!data.newBlocksSubscription) return

  data.newBlocksSubscription.unsubscribe((error, result) => {
    if (!result) {
      logger.error('Error while unsubscribing from new blocks.')
      logger.error(error?.message)
      logger.error(error?.stack)
    }

    logger.info('Unsubscribed from new blocks.')
    data.newBlocksSubscription = null
  })
}

const tryFetchTransactionsToUpdate = async (): Promise<string[]> => {
  return await tryAction<string[]>(fetchTransactionsToUpdate, RETRY_WAIT, MAX_RETRIES, (error) => {
    logger.error('Error while fetching transactions to update.')
    logger.error(error.message)
    logger.error(error.stack)
  })
}

const tryFetchTransactionReceipt = async (hash: string, data: EthereumData): Promise<Transaction> => {
  try {
    return await fetchTransactionReceipt(hash, data)
  } catch (error) {
    logger.error(`Error while fetching receipt for '${hash}'.`)
    logger.error(error.message)
    logger.error(error.stack)
    return null
  }
}

const fetchTransactionReceipts = async (data: EthereumData): Promise<void> => {
  if (!data.fetchingTransactionReceipts) return
  logger.info('Fetching transaction receipts...')

  const hashes = await tryFetchTransactionsToUpdate()
  logger.info(`${hashes.length} transaction(s) to update...`)

  const transactions = []
  if (hashes.length) {
    for (const hash of hashes) {
      if (!data.fetchingTransactionReceipts) return
      const transaction = await tryFetchTransactionReceipt(hash, data)

      if (transaction) {
        transactions.push(transaction)
      } else {
        logger.info(`Receipt not ready for transaction '${hash}'.`)
      }
    }

    if (transactions.length) {
      const updated = await trySaveTransactions(transactions)

      if (updated) logger.info(`${transactions.length} transaction(s) updated.`)
    } else {
      logger.info('No transaction to update.')
    }
  }

  if (!hashes.length || !transactions.length) {
    logger.debug(`Waiting ${TRANSACTION_RECEIPTS_FETCHING_INTERVAL / 1000} seconds...`)
    await sleep(TRANSACTION_RECEIPTS_FETCHING_INTERVAL)
  }

  fetchTransactionReceipts(data)
}

export const startFetchingTransactionReceipts = (data: EthereumData): void => {
  if (data.fetchingTransactionReceipts) return

  logger.info('Start fetching transaction receipts.')
  data.fetchingTransactionReceipts = true
  fetchTransactionReceipts(data)
}

export const stopFetchingTransactionReceipts = (data: EthereumData): void => {
  if (!data.fetchingTransactionReceipts) return

  data.fetchingTransactionReceipts = false
  logger.info('Stop fetching transaction receipts.')
}

const tryFetchTokenPrice = async (token: Token, data: EthereumData): Promise<boolean> => {
  try {
    await fetchTokenPrice(token, data)

    return true
  } catch (error) {
    logger.error(`Error while fetching price for ${token.label}.`)
    logger.error(error.message)
    logger.error(error.stack)

    return false
  }
}

const tryUpdateTokenPrice = async (token: Token): Promise<boolean> => {
  try {
    await updateTokenPrice(token)

    return true
  } catch (error) {
    logger.error(`Error while updating price for ${token.label}.`)
    logger.error(error.message)
    logger.error(error.stack)

    return false
  }
}

const fetchTokenPrices = async (data: EthereumData): Promise<void> => {
  if (!data.fetchingTokenPrices) return

  const tokens = data.contracts.filter((contract) => contract.type == ContractType.Token && contract.hash != WETH_HASH)

  logger.info(`Fetching prices for ${tokens.length} tokens...`)

  for (const token of tokens) {
    if (!data.fetchingTokenPrices) return

    const result = await tryFetchTokenPrice(token as Token, data)

    if (result) await tryUpdateTokenPrice(token as Token)
  }

  logger.info(`Token prices fetched.`)
  logger.debug(`Waiting ${TOKEN_PRICE_FETCHING_INTERVAL / 1000} seconds...`)
  await sleep(TOKEN_PRICE_FETCHING_INTERVAL)

  fetchTokenPrices(data)
}

export const startFetchingTokenPrices = (data: EthereumData): void => {
  if (data.fetchingTokenPrices) return

  logger.info('Start fetching token prices.')
  data.fetchingTokenPrices = true
  fetchTokenPrices(data)
}

export const stopFetchingTokenPrices = (data: EthereumData): void => {
  if (!data.fetchingTokenPrices) return

  data.fetchingTokenPrices = false
  logger.info('Stop fetching token prices.')
}

const tryFetchTransaction = async (hash: string, data: EthereumData): Promise<Transaction> => {
  return await tryAction<Transaction>(
    () => fetchPendingTransaction(hash, data),
    RETRY_WAIT,
    MAX_RETRIES,
    (error) => {
      logger.error(`Error while fetching transaction ${hash}.`)
      logger.error(error.message)
      logger.error(error.stack)
    },
    () => logger.trace(`Transaction ${hash} not available yet.`)
  )
}

const trySaveTransaction = async (transaction: Transaction): Promise<boolean> => {
  const result = await tryAction<boolean>(
    async () => {
      await saveTransaction(transaction)

      return true
    },
    RETRY_WAIT,
    MAX_RETRIES,
    (error) => {
      logger.error(`Error while saving transaction ${transaction.hash}.`)
      logger.error(error.message)
      logger.error(error.stack)
    }
  )

  return !!result
}

const onNewPendingTransaction = async (hash: string, data: EthereumData): Promise<void> => {
  logger.trace(`New pending transaction received: ${hash}.`)

  const transaction = await tryFetchTransaction(hash, data)

  if (!transaction || !isTransactionWatched(transaction, data.contracts)) return

  const result = await trySaveTransaction(transaction)

  if (result) logger.info(`Pending transaction added: ${transaction.hash}.`)
}

export const subscribeToPendingTransactions = (data: EthereumData): void => {
  if (data.pendingTransactionsSubscription) return

  data.pendingTransactionsSubscription = data.web3Pool.eth
    .subscribe('pendingTransactions')
    .on('connected', () => logger.info('Subscribed to pending transactions.'))
    .on('data', (hash) => onNewPendingTransaction(hash, data))
    .on('error', (error) => {
      logger.error('Error from pending transactions subscription.')
      logger.error(error?.message)
      logger.error(error?.stack)
    })
}

export const unsubscribeFromPendingTransactions = (data: EthereumData): void => {
  if (!data.pendingTransactionsSubscription) return

  data.pendingTransactionsSubscription.unsubscribe((error, result) => {
    if (!result) {
      logger.error('Error while unsubscribing from pending transactions.')
      logger.error(error?.message)
      logger.error(error?.stack)
    }

    logger.info('Unsubscribed from pending transactions.')
    data.pendingTransactionsSubscription = null
  })
}
