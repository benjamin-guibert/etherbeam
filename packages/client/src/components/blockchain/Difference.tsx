import React, { ReactElement } from 'react'
import numeral from 'numeral'
import { HistoryTime } from 'libraries/ethereum/types'
import Badge from '../Badge'
import './Difference.scss'

interface DifferenceProps {
  ratio: number
  time?: HistoryTime
}

const Difference = ({ ratio, time }: DifferenceProps): ReactElement => {
  const getColor = () => {
    if (ratio > 1.01) {
      return 'success'
    } else if (ratio < 0.99) {
      return 'danger'
    } else {
      return 'neutral'
    }
  }

  return (
    <Badge type={getColor()}>
      <span className="my-difference-prefix">{time.toUpperCase()}</span>
      <span className="my-amount">{numeral(ratio - 1).format('+0%')}</span>
    </Badge>
  )
}

export default Difference
