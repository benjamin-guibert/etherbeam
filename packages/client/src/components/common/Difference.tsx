import React, { ReactElement } from 'react'
import numeral from 'numeral'
import { Badge } from 'react-bootstrap'
import { HistoryTime } from 'libraries/ethereum/types'

interface DifferenceProps {
  ratio: number
  time?: HistoryTime
}

const Difference = ({ ratio, time }: DifferenceProps): ReactElement => {
  const getVariant = () => {
    if (ratio > 1.01) {
      return 'success'
    } else if (ratio < 0.99) {
      return 'danger'
    } else {
      return 'secondary'
    }
  }

  return (
    <Badge variant={getVariant()} className="text-monospace">
      {time && <span className="mr-2">{time.toUpperCase()}</span>}
      {numeral(ratio - 1).format('+0 %')}
    </Badge>
  )
}

export default Difference
