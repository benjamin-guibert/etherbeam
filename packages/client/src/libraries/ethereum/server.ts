import axios from 'axios'
import compact from 'lodash/compact'
import { HttpStatus, getPagination, PaginatedContent } from 'libraries/api'
import { Token, Transaction, USDC_HASH, WETH_HASH } from './types'
import { TokenData, TransactionData } from './serverTypes'
import { parseTokenData, parseTransactionData } from './serverParsers'

const TIMEOUT = 60 * 1000
const TRANSACTIONS_PER_PAGE = 20

const axiosClient = axios.create({
  baseURL: process.env.SERVER_API_HOST,
  headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
  timeout: TIMEOUT,
})

export const fetchTokens = async (): Promise<Token[]> => {
  const { data } = await axiosClient.get<TokenData[]>('tokens')
  const ignoredHashes = [WETH_HASH, USDC_HASH]

  return compact(data.map((token) => (!ignoredHashes.includes(token.address_hash) ? parseTokenData(token) : null)))
}

export const fetchToken = async (address: string, withList?: 'actions'): Promise<Token> => {
  const config = {
    validateStatus: (status: number) => [HttpStatus.Ok, HttpStatus.NotFound].includes(status),
  }

  const { data } = await axiosClient.get<TokenData>(`tokens/${address}?list=${withList || ''}`, config)

  return data ? parseTokenData(data) : null
}

export const fetchPendingTransactions = async (page: number): Promise<PaginatedContent<Transaction>> => {
  const response = await axiosClient.get<TransactionData[]>(
    `block_transactions/pending?items=${TRANSACTIONS_PER_PAGE}&page=${page}`
  )

  return {
    items: response.data.map((transaction) => parseTransactionData(transaction)),
    pagination: getPagination(response),
  }
}
