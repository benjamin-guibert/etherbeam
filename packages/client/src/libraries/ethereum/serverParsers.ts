import { BigNumber } from 'ethers'
import {
  TransactionData,
  TokenData,
  AddressData,
  TokenActionData,
  TransactionMethodParameterData,
  TransactionMethodData,
} from './serverTypes'
import {
  Transaction,
  Token,
  HistoryTime,
  Address,
  AddressType,
  TransactionMethodValueType,
  TransactionAction,
  TransactionActionDirection,
  TransactionMethodParameter,
  TransactionMethodParameterType,
  TransactionMethod,
} from './types'
import { createTransactionAddress, getEtherscanUrl } from './ethereum'
import orderBy from 'lodash/orderBy'

const parseBigNumberFromServer = (value: string): BigNumber => (value ? BigNumber.from(value.split('.')[0]) : null)

export const parseAddressData = ({ address_hash, address_type, label }: AddressData): Address => {
  return {
    hash: address_hash,
    type: address_type,
    label,
    url: getEtherscanUrl('address', address_hash),
  }
}

const parseHashData = (hash: string, address?: AddressData | TokenData): Address => {
  if (!address) {
    return {
      hash,
      type: AddressType.Unknown,
      url: getEtherscanUrl('address', hash),
    }
  }

  switch (address.address_type) {
    case AddressType.Token:
      return parseTokenData(address as TokenData)
    default:
      return parseAddressData(address)
  }
}

const parseTransactionMethodValueType = ({
  parameter_type,
  decimal_value,
  addresses,
}: TransactionMethodParameterData): TransactionMethodValueType => {
  switch (parameter_type) {
    case TransactionMethodParameterType.UnsignedInteger:
      return parseBigNumberFromServer(decimal_value)
    case TransactionMethodParameterType.Address:
      return addresses.length > 0 ? parseHashData(addresses[0].address_hash, addresses[0].address) : null
    case TransactionMethodParameterType.Addresses:
      return addresses ? addresses.map((address) => parseHashData(address.address_hash, address.address)) : null
    default:
      return null
  }
}

const parseTransactionMethodParameter = (parameter: TransactionMethodParameterData): TransactionMethodParameter => {
  const { name, parameter_type, raw_type } = parameter

  return {
    name,
    type: parameter_type,
    rawType: raw_type,
    value: parseTransactionMethodValueType(parameter),
  }
}

const parseTransactionMethodData = ({ contract, name, parameters }: TransactionMethodData): TransactionMethod => {
  return {
    contract: parseAddressData(contract),
    name,
    parameters: parameters.map((parameter) => parseTransactionMethodParameter(parameter)),
  }
}

export const parseTransactionData = ({
  transaction_hash,
  status,
  datetime,
  block_number,
  from_address_hash,
  from_address,
  to_address_hash,
  to_address,
  value,
  gas_unit_price,
  gas_limit,
  gas_used,
  gas_ratio,
  gas_fee,
  transaction_method_action,
  transaction_method_logs,
}: TransactionData): Transaction => {
  return {
    blockNumber: block_number,
    dateTime: datetime,
    address: createTransactionAddress(transaction_hash),
    status,
    from: from_address_hash ? parseHashData(from_address_hash, from_address) : null,
    to: to_address_hash ? parseHashData(to_address_hash, to_address) : null,
    value: parseBigNumberFromServer(value),
    gas: {
      unitPrice: parseBigNumberFromServer(gas_unit_price),
      limit: gas_limit,
      used: gas_used,
      ratio: gas_ratio,
      price: parseBigNumberFromServer(gas_fee),
    },
    action: transaction_method_action ? parseTransactionMethodData(transaction_method_action) : null,
    logs: transaction_method_logs?.map((log) => parseTransactionMethodData(log)),
  }
}

const parseTokenActionData = (
  {
    block_transaction,
    holder_address_hash,
    holder_address,
    index,
    action_type,
    from_address,
    from_address_hash,
    from_amount,
    to_address,
    to_address_hash,
    to_amount,
  }: TokenActionData,
  direction: TransactionActionDirection
): TransactionAction => {
  return {
    transaction: parseTransactionData(block_transaction),
    type: action_type,
    direction,
    index,
    holder: parseHashData(holder_address_hash, holder_address),
    fromAddress: from_address_hash ? parseHashData(from_address_hash, from_address) : null,
    fromAmount: parseBigNumberFromServer(from_amount),
    toAddress: to_address_hash ? parseHashData(to_address_hash, to_address) : null,
    toAmount: parseBigNumberFromServer(to_amount),
  }
}

export const parseTokenData = (data: TokenData): Token => {
  const {
    address_hash,
    name,
    description,
    symbol,
    decimals,
    price,
    price_1h,
    price_1h_ratio,
    price_1d,
    price_1d_ratio,
    price_1w,
    price_1w_ratio,
    price_1m,
    price_1m_ratio,
    price_1y,
    price_1y_ratio,
    website,
    whitepaper,
    github,
    linkedin,
    facebook,
    reddit,
    twitter,
    telegram,
    discord,
    from_transaction_actions,
    to_transaction_actions,
  } = data
  const priceHistory = []
  if (price_1h) {
    priceHistory.push({ time: HistoryTime.Hour, price: parseBigNumberFromServer(price_1h), ratio: price_1h_ratio })
  }
  if (price_1d) {
    priceHistory.push({ time: HistoryTime.Day, price: parseBigNumberFromServer(price_1d), ratio: price_1d_ratio })
  }
  if (price_1w) {
    priceHistory.push({ time: HistoryTime.Week, price: parseBigNumberFromServer(price_1w), ratio: price_1w_ratio })
  }
  if (price_1m) {
    priceHistory.push({ time: HistoryTime.Month, price: parseBigNumberFromServer(price_1m), ratio: price_1m_ratio })
  }
  if (price_1y) {
    priceHistory.push({ time: HistoryTime.Year, price: parseBigNumberFromServer(price_1y), ratio: price_1y_ratio })
  }

  const fromActions = from_transaction_actions
    ? from_transaction_actions.map((action) => parseTokenActionData(action, TransactionActionDirection.Sell))
    : []
  const toActions = to_transaction_actions
    ? to_transaction_actions.map((action) => parseTokenActionData(action, TransactionActionDirection.Buy))
    : []
  const actions =
    from_transaction_actions || to_transaction_actions
      ? orderBy([...fromActions, ...toActions], ['transaction.dateTime', 'index'], ['desc', 'desc'])
      : null

  return {
    ...parseAddressData(data),
    type: AddressType.Token,
    url: getEtherscanUrl('token', address_hash),
    name,
    description,
    symbol,
    decimals,
    price: price ? parseBigNumberFromServer(price) : null,
    priceHistory,
    website,
    whitepaper,
    github,
    linkedin,
    facebook,
    reddit,
    twitter,
    telegram,
    discord,
    actions,
  }
}
