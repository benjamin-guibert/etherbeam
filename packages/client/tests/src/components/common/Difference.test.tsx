import * as React from 'react'
import { mount, ReactWrapper } from 'enzyme'
import Difference from 'components/common/Difference'
import { HistoryTime } from 'libraries/ethereum/types'

describe('<Difference />', () => {
  let component: ReactWrapper

  describe('Hour', () => {
    beforeAll(() => {
      component = mount(<Difference time={HistoryTime.Hour} ratio={1} />)
    })

    it('should render', () => {
      const badge = component.find('Badge')
      expect(badge.prop('variant')).toEqual('secondary')
      expect(badge.find('span').at(1).prop('children')).toEqual('1H')
      expect(badge.prop('children')[1]).toEqual('+0 %')
    })

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Day', () => {
    beforeAll(() => {
      component = mount(<Difference time={HistoryTime.Day} ratio={1} />)
    })

    it('should render', () => {
      const badge = component.find('Badge')
      expect(badge.prop('variant')).toEqual('secondary')
      expect(badge.find('span').at(1).prop('children')).toEqual('1D')
      expect(badge.prop('children')[1]).toEqual('+0 %')
    })
  })

  describe('Week', () => {
    beforeAll(() => {
      component = mount(<Difference time={HistoryTime.Week} ratio={1} />)
    })

    it('should render', () => {
      const badge = component.find('Badge')
      expect(badge.prop('variant')).toEqual('secondary')
      expect(badge.find('span').at(1).prop('children')).toEqual('1W')
      expect(badge.prop('children')[1]).toEqual('+0 %')
    })
  })

  describe('Month', () => {
    beforeAll(() => {
      component = mount(<Difference time={HistoryTime.Month} ratio={1} />)
    })

    it('should render', () => {
      const badge = component.find('Badge')
      expect(badge.prop('variant')).toEqual('secondary')
      expect(badge.find('span').at(1).prop('children')).toEqual('1M')
      expect(badge.prop('children')[1]).toEqual('+0 %')
    })
  })

  describe('Year', () => {
    beforeAll(() => {
      component = mount(<Difference time={HistoryTime.Year} ratio={1} />)
    })

    it('should render', () => {
      const badge = component.find('Badge')
      expect(badge.prop('variant')).toEqual('secondary')
      expect(badge.find('span').at(1).prop('children')).toEqual('1Y')
      expect(badge.prop('children')[1]).toEqual('+0 %')
    })
  })

  describe('Positive', () => {
    beforeAll(() => {
      component = mount(<Difference ratio={1.3} />)
    })

    it('should render', () => {
      const badge = component.find('Badge')
      expect(badge.prop('variant')).toEqual('success')
      expect(badge.prop('children')[1]).toEqual('+30 %')
    })
  })

  describe('Negative', () => {
    beforeAll(() => {
      component = mount(<Difference ratio={0.7} />)
    })

    it('should render', () => {
      const badge = component.find('Badge')
      expect(badge.prop('variant')).toEqual('danger')
      expect(badge.prop('children')[1]).toEqual('-30 %')
    })
  })
})
