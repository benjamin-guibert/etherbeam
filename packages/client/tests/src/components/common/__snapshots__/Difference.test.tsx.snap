// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Difference /> Hour should match snapshot 1`] = `
<span
  class="text-monospace badge badge-secondary"
>
  <span
    class="mr-2"
  >
    1H
  </span>
  +0 %
</span>
`;
